src/client/opamTreeCommand.ml

```diff
@@ -399,24 +402,25 @@ let simulate_new_state tog st universe install names =
          (OpamSwitchState.unavailable_reason st) cs);
     OpamStd.Sys.exit_because `No_solution
 
-let dry_install tog st universe install =
-  simulate_new_state tog st universe install
-    (OpamPackage.Name.Set.of_list (List.map fst install))
-
 let run st tog ?no_constraint mode filter atoms =
   let open OpamPackage.Set.Op in
-  let select, missing =
-    List.fold_left (fun (select, missing) atom ->
+  let select, missing, requested =
+    List.fold_left (fun (select, missing, requested) atom ->
```

Comment:
Instead of computing requested here, what do you think of computing the universe conditionally (regarding missing status):
```
  let st, universe =
    let universe =
      let requested =
        OpamFormula.packages_of_atoms
          (if missing = [] then st.installed else st.packages) atoms
      in
      get_universe tog requested st
    in
```

Reply:
I like the idea. Added in https://github.com/ocaml/opam/pull/5919/commits/36cd0528f8f704489e8b215834cda738007295e1